File: frontend/src/api/enhancedSearch.ts
import client from './client';

export interface SearchFilters {
  genre?: string;
  style?: string;
  tone?: string;
  target_audience?: string;
  status?: string;
  is_completed?: boolean;
  min_word_count?: number;
  max_word_count?: number;
}

export interface SearchResult {
  id: number;
  title: string;
  genre?: string;
  style?: string;
  tone?: string;
  target_audience?: string;
  total_word_count: number;
  status: string;
  created_at: string | null;
  author?: string;
  cover_image_url?: string;
}

export interface SearchResponse {
  results: SearchResult[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface SearchSuggestion {
  type: 'title' | 'genre';
  text: string;
}

export interface SuggestionsResponse {
  suggestions: SearchSuggestion[];
  query: string;
}

export interface SearchStats {
  genres: Array<{ name: string; count: number }>;
  styles: Array<{ name: string; count: number }>;
  target_audiences: Array<{ name: string; count: number }>;
}

export const enhancedSearchApi = {
  // Advanced search with filters (POST)
  advancedSearch: (data: {
    query: string;
    filters?: SearchFilters;
    page?: number;
    page_size?: number;
  }) =>
    client.post<SearchResponse>('/search/novels', {
      query: data.query,
      ...data.filters,
    }, {
      params: {
        page: data.page || 1,
        page_size: data.page_size || 20,
      },
    }),

  // Simple search (GET)
  simpleSearch: (params: {
    q: string;
    genre?: string;
    style?: string;
    page?: number;
    page_size?: number;
  }) =>
    client.get<SearchResponse>('/search/novels', { params }),

  // Get autocomplete suggestions
  getSuggestions: (query: string, limit?: number) =>
    client.get<SuggestionsResponse>('/search/suggestions', {
      params: {
        q: query,
        ...(limit ? { limit } : {}),
      },
    }),

  // Get search filter statistics
  getStats: () =>
    client.get<SearchStats>('/search/stats'),
};

File: backend/app/routers/enhanced_search.py
from typing import Optional
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.services.enhanced_search_service import enhanced_search_service

router = APIRouter(prefix="/search", tags=["search"])


class SearchRequest(BaseModel):
    query: str
    genre: Optional[str] = None
    style: Optional[str] = None
    tone: Optional[str] = None
    target_audience: Optional[str] = None
    status: Optional[str] = None
    is_completed: Optional[bool] = None
    min_word_count: Optional[int] = None
    max_word_count: Optional[int] = None


@router.post("/novels")
async def advanced_search(
    data: SearchRequest,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    """Advanced novel search with multiple filters."""
    filters = {
        "genre": data.genre,
        "style": data.style,
        "tone": data.tone,
        "target_audience": data.target_audience,
        "status": data.status,
        "is_completed": data.is_completed,
        "min_word_count": data.min_word_count,
        "max_word_count": data.max_word_count,
    }
    
    result = await enhanced_search_service.search_novels(
        db=db,
        query=data.query,
        filters=filters,
        page=page,
        page_size=page_size,
    )
    return result


@router.get("/novels")
async def search_novels_get(
    q: str,
    genre: Optional[str] = None,
    style: Optional[str] = None,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    """Simple GET search endpoint."""
    filters = {
        "genre": genre,
        "style": style,
    }
    
    result = await enhanced_search_service.search_novels(
        db=db,
        query=q,
        filters=filters,
        page=page,
        page_size=page_size,
    )
    return result


@router.get("/suggestions")
async def get_suggestions(
    q: str,
    limit: int = Query(default=10, ge=1, le=20),
    db: Session = Depends(get_db),
):
    """Get autocomplete search suggestions."""
    result = await enhanced_search_service.get_search_suggestions(
        db=db,
        partial_query=q,
        limit=limit,
    )
    return result


@router.get("/stats")
async def get_search_stats(
    db: Session = Depends(get_db),
):
    """Get search filter statistics (genres, styles, etc.)."""
    result = await enhanced_search_service.get_search_stats(db=db)
    return result
